src/microcalibrate/reweight.py

```diff
@@ -169,11 +178,106 @@ def dropout_weights(weights: torch.Tensor, p: float) -> torch.Tensor:
+        # NOTE: Results are pretty sensitve to learning rates
+        # optimizer breaks down somewhere near .005, does better at above .1
+        optimizer = torch.optim.Adam(
+            [weights] + list(gates.parameters()), lr=0.2
+        )
+        start_loss = None
+
+        loss_over_epochs_sparse = []
+        estimates_over_epochs_sparse = []
+        pct_close_over_epochs_sparse = []
+        epochs_sparse = []
+
+        iterator = tqdm(
+            range(epochs * 2), desc="Sparse reweighting progress", unit="epoch"
+        )  # lower learning rate, harder optimization
+
+        for i in iterator:
+            optimizer.zero_grad()
+            weights_ = dropout_weights(weights, dropout_rate)
+            masked = torch.exp(weights_) * gates()
+            estimate = estimate_function(masked)
+            l_main = loss(estimate, targets, normalization_factor)
+            l = l_main + l0_lambda * gates.get_penalty()
+            close = pct_close(estimate, targets)
+            if i % tracking_n / 2 == 0:
+                epochs_sparse.append(i)
+                loss_over_epochs_sparse.append(l.item())
+                pct_close_over_epochs_sparse.append(close)
+                estimates_over_epochs_sparse.append(
+                    estimate.detach().cpu().numpy()
+                )
+
+                logger.info(
+                    f"Within 10% from targets in sparse calibration: {close:.2%} \n"
+                )
+
+                if len(loss_over_epochs_sparse) > 1:
+                    loss_change = loss_over_epochs_sparse[-2] - l.item()
+                    logger.info(
+                        f"Epoch {i:4d}: Loss = {l.item():.6f}, "
+                        f"Change = {loss_change:.6f} "
+                        f"({'improving' if loss_change > 0 else 'worsening'})"
+                    )
+            if start_loss is None:
+                start_loss = l.item()
+            loss_rel_change = (l.item() - start_loss) / start_loss
+            l.backward()
+            iterator.set_postfix(
+                {"loss": l.item(), "loss_rel_change": loss_rel_change}
+            )
+            optimizer.step()
+
+        gates.eval()
+        final_weights_sparse = (
+            (torch.exp(weights) * gates()).detach().cpu().numpy()
+        )
+
+        tracker_dict_sparse = {
+            "epochs": epochs_sparse,
+            "loss": loss_over_epochs_sparse,
+            "estimates": estimates_over_epochs_sparse,
+        }
+
+        sparse_performance_df = log_performance_over_epochs(
+            tracker_dict_sparse,
+            targets,
+            target_names,
+            excluded_targets,
+            excluded_target_data,
+        )
+
+        if csv_path:
+            # Create directory if it doesn't exist
+            csv_path = Path(str(csv_path).replace(".csv", "_sparse.csv"))
+            csv_path.parent.mkdir(parents=True, exist_ok=True)
+            sparse_performance_df.to_csv(csv_path, index=True)
```

Comment:
I'm sure this functions as a perfectly complete log. I will say that I've benefitted from seeing the output of `print_reweighting_diagnostics` in loss.py from policyengine-us-data. Up to you whether you want to bring it in.